[✨feat] Color(Scrap) 구현

jsoonworld opened the pull request:
# 📄 Work Description  
Scrap 도메인에서 사용되는 색상(Color)을 enum으로 정의하고,  
label 기반의 조회 기능 및 HEX 코드 반환 기능을 구현했습니다.  
또한 유효하지 않은 label 입력 시 적절한 예외가 발생하도록 `ScrapException`과 `ScrapErrorCode`를 함께 설계하였습니다.  
모든 기능은 테스트 코드로 검증을 완료했습니다 😊

---

# 💭 Thoughts  
이번 작업에서 가장 고민했던 부분은 **조회 실패 시 어떤 방식으로 예외를 처리할지**였습니다.  
`Color` enum은 단순한 값 정의를 넘어 도메인적인 의미도 가지기 때문에,  
조회 실패를 일반적인 IllegalArgumentException이 아니라,  
**스크랩 도메인 전용 예외로 명확히 구분하는 방향**이 더 낫겠다고 판단했어요.

또 하나는 `entries.firstOrNull { it.label == label }`처럼  
컬렉션 캐싱 없이도 가독성 있게 작성할 수 있는 enum 활용 방식에 대해 고민해봤습니다.

혹시 이 enum 구조나 예외 처리 방식에서  
불필요하거나 과한 부분, 혹은 더 명확하게 다듬을 수 있는 부분이 있다면 편하게 말씀해 주세요!  
더 좋은 방향이 있다면 꼭 공유받고 싶어요 🙏

---

# ✅ Testing Result  

![스크린샷 2025-05-17 오후 8 56 26](https://github.com/user-attachments/assets/9d97b247-78ef-4870-aaa1-ddf32a3d071e)

---

# 🗂 Related Issue  
- closed #18

jsoonworld committed:
feat(Color) : 색상 enum 정의 및 변환 기능 구현

스크랩 색상을 표현하는 Color enum을 정의하고,
label 기반 조회 및 HEX 문자열 반환 기능을 추가했습니다.

jsoonworld committed:
feat(ColorTest) : Color enum 기능 검증 테스트 추가

label 기반 Color 매핑 및 HEX 문자열 반환 기능에 대한 단위 테스트를 작성했습니다.

jsoonworld committed:
feat(ScrapErrorCode) : 스크랩 도메인 전용 에러 코드 정의

스크랩 색상 유효성 검증 실패 시 사용할 INVALID_COLOR 에러 코드를 추가했습니다.

jsoonworld committed:
feat(ScrapException) : 스크랩 도메인 전용 예외 클래스 구현

ScrapErrorCode와 함께 사용하는 RuntimeException 기반의 커스텀 예외를 정의했습니다.

jsoonworld committed:
feat(ExceptionHandler) : ScrapException 전용 핸들러 추가

스크랩 도메인 예외를 처리하는 핸들러를 추가하여 적절한 HTTP 상태와 메시지를 반환하도록 구성했습니다.

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/scrap/Color.kt:
```diff
@@ -0,0 +1,26 @@
+package com.terning.server.kotlin.domain.scrap
+
+enum class Color(
+    val label: String,
+    private val hexCode: String,
+) {
```
엄청 꼼꼼하시네요..! 
Color라는 이넘 클래스를 사용할 때 label로만 접근하도록 hexCode는 private으로 둔 거 맞을까요?
한 수 배워갑니다..

jsoonworld replied:
꼼꼼하게 봐주셔서 감사합니다! ✨

말씀해주신 것처럼 hexCode를 private으로 둔 이유는 외부에서 직접 접근하기보다는, toHexString() 메서드를 통해서만 접근하도록 의도한 설계예요.
이렇게 하면 `hexCode의 표현 방식(예: # 붙이기)`을 이 내부 메서드에서 통제할 수 있어서, 사용하는 입장에선 항상 일관된 값을 얻게 되거든요!

`label`은 사용자 입력 등을 바탕으로 매핑하거나 보여줄 때 자주 활용되기 때문에, 공개적으로 열어두고 있고요.
이런 방식으로 역할을 명확히 나누면, 의도치 않게 잘못된 값이 노출되거나 사용되는 걸 막을 수 있어서 개인적으로 선호하는 방식이에요 😊

작은 부분까지 세심하게 봐주셔서 정말 감사드려요. 덕분에 더 좋은 논의가 되는 것 같아요 🙌

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/scrap/Color.kt:
```diff
@@ -0,0 +1,26 @@
+
+    fun toHexString(): String = "#$hexCode"
+
+    companion object {
+        fun from(label: String): Color =
+            entries.firstOrNull { it.label == label }
+                ?: throw ScrapException(ScrapErrorCode.INVALID_COLOR)
+    }
```
이렇게 작성해 주신 부분이 고민이셨다는 말씀이시죠...?
저는 괜찮은 방법인 것 같습니다!
보통 enum의 특징을 잘 살린 entries를 사용하곤 하더라구요

jsoonworld replied:
네 맞아요! 말씀해주신 부분에서 어떤 방식이 더 좋을지 살짝 고민이 있었는데,
entries를 활용하는 방식이 enum의 장점을 살리면서도 깔끔하게 매핑할 수 있어서 이 방법을 선택했어요.
특히 label처럼 사용자 입력 기반의 값과 enum을 연결할 때, entries.firstOrNull은 정말 유용한 것 같아요 🙌

괜찮은 방법이라고 말씀해주셔서 더 자신감이 생깁니다!
늘 꼼꼼하게 봐주셔서 감사해요 ✨

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/scrap/ScrapErrorCode.kt:
```diff
@@ -0,0 +1,10 @@
+enum class ScrapErrorCode(
+    val status: HttpStatus,
+    val message: String,
+) {
+    INVALID_COLOR(HttpStatus.BAD_REQUEST, "유효하지 않은 스크랩 색상입니다."),
+}
```
저는 보통 인자가 2개 이상 들어가면 명확하게 하기 위해 named arguments 형식으로 넣어주곤 하는데요..! 이렇게 작성해 주는 건 어떤가요..?!
```suggestion
) {
    INVALID_COLOR(status = HttpStatus.BAD_REQUEST, message = "유효하지 않은 스크랩 색상입니다."),
}
```

jsoonworld replied:
피드백 감사합니다! 🙌  
`named arguments`에 대해 제안해주신 부분도 정말 공감돼요.  
다만 이번 `enum`에서는 이미 프로퍼티 이름이 선언되어 있어서, IDE나 컴파일 타임에서도 인자 순서나 타입 오류를 잘 잡아주다 보니,  
개인적으로는 꼭 명시적으로 이름을 붙이지 않아도 괜찮지 않을까? 하는 생각이 있었어요.

그리고 내부적으로 다른 enum들도 인자 수가 1개에서 2개, 3개로 점차 늘어나고 있는데,  
이럴 경우 오히려 일부만 `named`로 작성하게 되면 스타일이 통일되지 않아 보일 수도 있다는 점도 조금 고민이 됐어요 😅

그래서 이번에는 기존 흐름을 유지하는 쪽으로 생각해봤는데, 어떻게 생각하시나요!?
이 부분은 컨벤션을 함께 맞춰가면 더 좋을 것 같아서 의견 여쭤봅니다! 😊

leeeyubin replied:
이 부분은 컨벤션을 맞추는 게 좋을 것 같네요!

`ScrapErrorCode`의 경우 두 프로퍼티가 타입이 다르기 때문에 오류가 나지 않지만, 코드를 작성하다보면 타입이 같은 여러 인자를 넣어줘야 하는 경우도 있기 때문에 `named arguments`로 작성해 주는 게 좋을 것 같아요..!

인자수가 2개 혹은 3개 이상일 때만 작성을 해 주거나 일관된 코드가 좋을 것 같다면 전부 작성해줘도 될 것 같다는 생각입니다!

jsoonworld replied:
피드백 감사합니다! 😊
예외 코드에서만 사용된다고 생각해서 positional arguments로 작성했었는데, 말씀해주신 대로 다른 곳에서도 사용될 가능성을 고려하면 named arguments로 작성하는 게 더 명확하고 안전하겠다는 생각이 들었습니다!

특히 타입이 동일한 인자가 늘어날수록 실수할 여지도 생기니, 이 부분은 저희 팀 컨벤션으로 잡아두는 게 좋을 것 같네요.
이슈로 등록해서 ScrapErrorCode뿐만 아니라 관련된 enum들도 함께 정리해보겠습니다. 

피드백 감사드려요! 🙏

leeeyubin reviewed (approved):
테스트 코드까지 꼼꼼하게 작성해 주셔서 편안하게 읽었습니다 ㅎㅎ 너무 수고하셨어요!!👍

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/domain/scrap/Color.kt:
```diff
@@ -0,0 +1,26 @@
+package com.terning.server.kotlin.domain.scrap
+
+enum class Color(
+    val label: String,
```
추가로 이 부분에서 `label`이라는 이름이 약간 모호하게 느껴질 수 있을 것 같아서,  
`color`로 변수명을 바꾸는 게 더 명확하겠다는 생각이 들었어요.  

예를 들어, `Color.from(color)`처럼 표현되면 직관적으로 "이 color 이름에 해당하는 enum을 찾는구나" 하고 이해하기 쉬울 것 같아서요!  
작은 네이밍이지만 읽는 사람 입장에서 조금 더 명확한 흐름이 되지 않을까 해서 한 번 반영해보려고 합니다 🙌  
혹시 이 부분에 대해서도 의견 있으시면 편하게 말씀주세요! 😊

jsoonworld committed:
refactor(Color) : color 프로퍼티명을 label로 변경

- 기존 color 명칭이 클래스명(Color)과 겹쳐 가독성이 다소 떨어지는 점을 고려해 label로 변경했습니다.
- enum 내 name은 내장 프로퍼티와 충돌 가능성이 있어 사용하지 않았습니다.
- 사용자 입력값과 매핑 시 의미를 명확히 전달하면서도, 코드 일관성을 유지할 수 있도록 개선했습니다.
